### PR TITLE
Fix path transition test

### PR DESCRIPTION
**Pull request by WildFireFlum**
This branch adds a test for the transition from the FAST_OPTIMISTIC commit path to the FAST_WITH_THRESHOLD path.
It changes the way metrics are counted as they previously did not account for three commit paths.
The full (trinary) path information is not propagated through the consensus protocol, but is propagated through the execution path, and is therefore used.
This branch also adds optional prints and seeded test runs for easier debugging process.

### CODE REVIEW

**Comment by teoparvanov** [comment]
Hi @WildFireFlum , can you please provide some description? What is the motivation of this change, what is a summary of the proposed changes?

**Review by evdzhurov — COMMENTED** [review]
I guess the main purpose of the PR is to add a test for the transition between the two fast path modes - optimistic and with threshold, however I have couple of observations:
* There are a lot of changes here, some of which I'm not sure if they are related to fixing the path transition test, like the addition of a random seed.
*  The PR adds a new method wait_for_consensus_path to detect if some execution path transition happened.  Since we already have _wait_for_consensus_path_to_be_prevalent how are we going to differentiate between the two if someone needs to decide what to use in a test? Is wait_for_consensus_path meant to detect the switch but doesn't guarantee prevalence? Further, wait_for_consensus_path says it counts operation executions but gathers metrics per pre-prepare which may contain a number of operations.
* Lastly, the change to how metrics are generated is now solely based on the firstPath parameter of a pre-prepare, which is the desired commit path but not necessarily the commited path since the slow path will begin competing with the fast path if consensus slows down for some sequence. This might still work in the case when we want to detect a switch in the state of the controller (what is the desired commit path) but I'm not sure how it will affect the stability of the alternative logic in _wait_for_consensus_path_to_be_prevalent, which wasn't ideal to begin with.

**Comment by WildFireFlum** [comment]
> Hi @WildFireFlum , can you please provide some description to the PR? What is the motivation of this change, what is a summary of the proposed changes?

I added a description.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -4696,6 +4680,7 @@ void ReplicaImp::setConflictDetectionBlockId(const ClientRequestMsg &clientReqMs
 // have some committed seq numbers)
 // TODO(GG): move the "logic related to requestMissingInfo" to the caller of this method (e.g., by returning a value
 // to the caller)
+// Pushes execution requests to another threa
```
Comment (yontyon) [inline bftengine/src/bftengine/ReplicaImp.cpp:4683]: Nit: threa -> thread

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -5184,6 +5169,26 @@ void ReplicaImp::finalizeExecution() {
...
+      metric_fast_threshold_path_count_++;
+      break;
+    case CommitPath::SLOW:
+      metric_total_slowPath_++;
+      metric_total_slowPath_requests_ += numOfRequests;
+      break;
+    case CommitPath::NA:
+      break;
+  }
+  LOG_INFO(GL, "Commit path metric update " << KVLOG(std::to_string((int8_t)executionPath)));
```
Comment (yontyon) [inline bftengine/src/bftengine/ReplicaImp.cpp:5189]: move to DEBUG, we have the commit path as part of the MDC

File: bftengine/src/preprocessor/GlobalData.cpp [hunk 3]
```diff
@@ -4,7 +4,4 @@ namespace preprocessor {
 
 std::atomic_uint64_t GlobalData::current_block_id = 0;
 std::atomic_uint64_t GlobalData::block_delta = 0;
-std::atomic_uint64_t GlobalData::delta_factor = 60;
-std::atomic_bool GlobalData::increment_step = false;
-std::atomic_bool GlobalData::decrement_step = false;
```
Comment (yontyon) [inline bftengine/src/preprocessor/GlobalData.cpp:9]: ???

File: bftengine/src/preprocessor/GlobalData.hpp [hunk 4]
```diff
@@ -6,9 +6,6 @@ struct GlobalData {
   // Used for the conflict detection optimization.
   static std::atomic_uint64_t current_block_id;
   static std::atomic_uint64_t block_delta;
-  static std::atomic_uint64_t delta_factor;
-  static std::atomic_bool increment_step;
-  static std::atomic_bool decrement_step;
```
Comment (yontyon) [inline bftengine/src/preprocessor/GlobalData.hpp:11]: ???

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 5]
```diff
@@ -1605,7 +1605,6 @@ OperationResult PreProcessor::launchReqPreProcessing(const PreProcessRequestMsgS
              "Primary block [" << preProcessReqMsg->primaryBlockId()
                                << "] is too advanced for conflict detection optimization, replica block id ["
                                << GlobalData::current_block_id << "] delta [" << GlobalData::block_delta << "]");
-    GlobalData::increment_step = true;
```
Comment (yontyon) [inline bftengine/src/preprocessor/PreProcessor.cpp:1608]: ???

**Review by yontyon — COMMENTED** [review]
LGTM!
One comment, I would like to have a separate PR for the Apollo improvements, is it possible?

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 2 above)
Comment (WildFireFlum) [reply bftengine/src/bftengine/ReplicaImp.cpp:5189]: This wasn't supposed to be committed, I'll remove it.

File: bftengine/src/preprocessor/GlobalData.cpp (hunk 3 above)
Comment (WildFireFlum) [reply bftengine/src/preprocessor/GlobalData.cpp:9]: These are not used anywhere in the code.

File: bftengine/src/preprocessor/GlobalData.hpp (hunk 4 above)
Comment (WildFireFlum) [reply bftengine/src/preprocessor/GlobalData.hpp:11]: Unused

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 5 above)
Comment (WildFireFlum) [reply bftengine/src/preprocessor/PreProcessor.cpp:1608]: unused

File: bftengine/src/preprocessor/GlobalData.cpp (hunk 3 above)
Comment (yontyon) [reply bftengine/src/preprocessor/GlobalData.cpp:9]: so move them to the other PR (the one with the overall fixes and changes)

**Pull request closed without merging**